File: src/lib/jobs/safe-fetch.ts
import dns from "node:dns/promises";
import net from "node:net";
import { isIP } from "node:net";

const MAX_BYTES = 1_500_000;
const TIMEOUT_MS = 10_000;
const MAX_REDIRECTS = 5;

export type SafeFetchResult = {
  ok: boolean;
  status: number;
  finalUrl: string;
  bodyText: string;
  error?: string;
};

function isPrivateIp(ip: string): boolean {
  if (net.isIPv4(ip)) {
    const parts = ip.split(".").map(Number);
    const [a, b] = parts;
    if (a === 10) return true;
    if (a === 127) return true;
    if (a === 0) return true;
    if (a === 169 && b === 254) return true;
    if (a === 172 && b >= 16 && b <= 31) return true;
    if (a === 192 && b === 168) return true;
    return false;
  }

  if (net.isIPv6(ip)) {
    const normalized = ip.toLowerCase();
    if (normalized === "::1") return true;
    if (normalized.startsWith("fc") || normalized.startsWith("fd")) return true;
    if (normalized.startsWith("fe80")) return true;
    return false;
  }

  return true;
}

async function assertPublicHostname(hostname: string) {
  if (hostname === "localhost" || hostname.endsWith(".local")) {
    throw new Error("Private or local hostnames are not allowed");
  }

  if (isIP(hostname)) {
    if (isPrivateIp(hostname)) {
      throw new Error("Private IP addresses are not allowed");
    }
    return;
  }

  const records = await dns.lookup(hostname, { all: true });
  if (records.length === 0) {
    throw new Error("Hostname could not be resolved");
  }

  for (const record of records) {
    if (isPrivateIp(record.address)) {
      throw new Error("Hostname resolves to a private address");
    }
  }
}

export async function safeFetch(
  rawUrl: string,
  options: { method?: "GET" | "HEAD"; accept?: string } = {},
): Promise<SafeFetchResult> {
  let current = rawUrl;

  try {
    for (let redirectCount = 0; redirectCount <= MAX_REDIRECTS; redirectCount++) {
      const url = new URL(current);
      if (url.protocol !== "http:" && url.protocol !== "https:") {
        return {
          ok: false,
          status: 0,
          finalUrl: current,
          bodyText: "",
          error: "Only HTTP and HTTPS URLs are allowed",
        };
      }

      await assertPublicHostname(url.hostname);

      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS);

      try {
        const response = await fetch(url.toString(), {
          method: options.method ?? "GET",
          redirect: "manual",
          signal: controller.signal,
          headers: {
            "User-Agent": "JobTrackerLocal/1.0",
            Accept: options.accept ?? "text/html,application/json,*/*",
          },
        });

        if ([301, 302, 303, 307, 308].includes(response.status)) {
          const location = response.headers.get("location");
          if (!location) {
            return {
              ok: false,
              status: response.status,
              finalUrl: current,
              bodyText: "",
              error: "Redirect missing Location header",
            };
          }
          current = new URL(location, current).toString();
          continue;
        }

        const contentLength = Number(response.headers.get("content-length") ?? "0");
        if (contentLength > MAX_BYTES) {
          return {
            ok: false,
            status: response.status,
            finalUrl: current,
            bodyText: "",
            error: "Response exceeds size limit",
          };
        }

        const buffer = Buffer.from(await response.arrayBuffer());
        if (buffer.byteLength > MAX_BYTES) {
          return {
            ok: false,
            status: response.status,
            finalUrl: current,
            bodyText: "",
            error: "Response exceeds size limit",
          };
        }

        return {
          ok: response.ok,
          status: response.status,
          finalUrl: current,
          bodyText: buffer.toString("utf8"),
        };
      } finally {
        clearTimeout(timeout);
      }
    }

    return {
      ok: false,
      status: 0,
      finalUrl: current,
      bodyText: "",
      error: "Too many redirects",
    };
  } catch (error) {
    const message =
      error instanceof Error
        ? error.name === "AbortError"
          ? "Request timed out"
          : error.message
        : "Request failed";

    return {
      ok: false,
      status: 0,
      finalUrl: current,
      bodyText: "",
      error: message,
    };
  }
}

export function extractHtmlTitle(html: string): string | null {
  const match = html.match(/<title[^>]*>([^<]*)<\/title>/i);
  if (!match?.[1]) {
    return null;
  }
  return match[1].replace(/\s+/g, " ").trim() || null;
}

export function looksLikeClosedPosting(html: string, status: number): boolean {
  if (status === 404 || status === 410) {
    return true;
  }

  const lower = html.toLowerCase();
  const closedSignals = [
    "no longer accepting applications",
    "job is closed",
    "this job has expired",
    "position has been filled",
    "this posting is no longer available",
    "sorry, this job is no longer available",
  ];

  return closedSignals.some((signal) => lower.includes(signal));
}

File: src-tauri/src/jobs/safe_fetch.rs
use std::net::{IpAddr, Ipv6Addr};
use std::time::Duration;

use dns_lookup::lookup_host;
use reqwest::redirect::Policy;
use reqwest::{Client, Method, StatusCode};
use url::Url;

const MAX_BYTES: usize = 1_500_000;
const TIMEOUT_MS: u64 = 10_000;
const MAX_REDIRECTS: usize = 5;

#[derive(Debug, Clone, serde::Serialize)]
pub struct SafeFetchResult {
    pub ok: bool,
    pub status: u16,
    pub final_url: String,
    pub body_text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

fn is_private_ip(ip: IpAddr) -> bool {
    match ip {
        IpAddr::V4(v4) => {
            let octets = v4.octets();
            let a = octets[0];
            let b = octets[1];
            a == 10
                || a == 127
                || a == 0
                || (a == 169 && b == 254)
                || (a == 172 && (16..=31).contains(&b))
                || (a == 192 && b == 168)
        }
        IpAddr::V6(v6) => {
            if v6 == Ipv6Addr::LOCALHOST {
                return true;
            }
            let s = v6.to_string().to_lowercase();
            s.starts_with("fc") || s.starts_with("fd") || s.starts_with("fe80")
        }
    }
}

fn assert_public_hostname(hostname: &str) -> Result<(), String> {
    if hostname == "localhost" || hostname.ends_with(".local") {
        return Err("Private or local hostnames are not allowed".into());
    }

    if let Ok(ip) = hostname.parse::<IpAddr>() {
        if is_private_ip(ip) {
            return Err("Private IP addresses are not allowed".into());
        }
        return Ok(());
    }

    let records = lookup_host(hostname).map_err(|e| e.to_string())?;
    if records.is_empty() {
        return Err("Hostname could not be resolved".into());
    }
    for addr in records {
        if is_private_ip(addr) {
            return Err("Hostname resolves to a private address".into());
        }
    }
    Ok(())
}

pub async fn safe_fetch(
    raw_url: &str,
    method: Option<&str>,
    accept: Option<&str>,
) -> SafeFetchResult {
    let method = match method.unwrap_or("GET").to_uppercase().as_str() {
        "HEAD" => Method::HEAD,
        _ => Method::GET,
    };
    let accept = accept.unwrap_or("text/html,application/json,*/*");
    let mut current = raw_url.to_string();

    let client = match Client::builder()
        .redirect(Policy::none())
        .timeout(Duration::from_millis(TIMEOUT_MS))
        .user_agent("JobTrackerLocal/1.0")
        .build()
    {
        Ok(c) => c,
        Err(e) => {
            return SafeFetchResult {
                ok: false,
                status: 0,
                final_url: current,
                body_text: String::new(),
                error: Some(e.to_string()),
            };
        }
    };

    for _ in 0..=MAX_REDIRECTS {
        let url = match Url::parse(&current) {
            Ok(u) => u,
            Err(e) => {
                return SafeFetchResult {
                    ok: false,
                    status: 0,
                    final_url: current,
                    body_text: String::new(),
                    error: Some(e.to_string()),
                };
            }
        };

        if url.scheme() != "http" && url.scheme() != "https" {
            return SafeFetchResult {
                ok: false,
                status: 0,
                final_url: current,
                body_text: String::new(),
                error: Some("Only HTTP and HTTPS URLs are allowed".into()),
            };
        }

        let host = url.host_str().unwrap_or("").to_string();
        if let Err(e) = assert_public_hostname(&host) {
            return SafeFetchResult {
                ok: false,
                status: 0,
                final_url: current,
                body_text: String::new(),
                error: Some(e),
            };
        }

        let response = match client
            .request(method.clone(), url.clone())
            .header("Accept", accept)
            .send()
            .await
        {
            Ok(r) => r,
            Err(e) => {
                let message = if e.is_timeout() {
                    "Request timed out".to_string()
                } else {
                    e.to_string()
                };
                return SafeFetchResult {
                    ok: false,
                    status: 0,
                    final_url: current,
                    body_text: String::new(),
                    error: Some(message),
                };
            }
        };

        let status = response.status();
        if matches!(
            status,
            StatusCode::MOVED_PERMANENTLY
                | StatusCode::FOUND
                | StatusCode::SEE_OTHER
                | StatusCode::TEMPORARY_REDIRECT
                | StatusCode::PERMANENT_REDIRECT
        ) {
            let location = match response.headers().get("location").and_then(|v| v.to_str().ok()) {
                Some(l) => l.to_string(),
                None => {
                    return SafeFetchResult {
                        ok: false,
                        status: status.as_u16(),
                        final_url: current,
                        body_text: String::new(),
                        error: Some("Redirect missing Location header".into()),
                    };
                }
            };
            current = match Url::parse(&location)
                .or_else(|_| url.join(&location))
                .map(|u| u.to_string())
            {
                Ok(u) => u,
                Err(e) => {
                    return SafeFetchResult {
                        ok: false,
                        status: status.as_u16(),
                        final_url: current,
                        body_text: String::new(),
                        error: Some(e.to_string()),
                    };
                }
            };
            continue;
        }

        if let Some(len) = response.content_length() {
            if len as usize > MAX_BYTES {
                return SafeFetchResult {
                    ok: false,
                    status: status.as_u16(),
                    final_url: current,
                    body_text: String::new(),
                    error: Some("Response exceeds size limit".into()),
                };
            }
        }

        let bytes = match response.bytes().await {
            Ok(b) => b,
            Err(e) => {
                return SafeFetchResult {
                    ok: false,
                    status: status.as_u16(),
                    final_url: current,
                    body_text: String::new(),
                    error: Some(e.to_string()),
                };
            }
        };

        if bytes.len() > MAX_BYTES {
            return SafeFetchResult {
                ok: false,
                status: status.as_u16(),
                final_url: current,
                body_text: String::new(),
                error: Some("Response exceeds size limit".into()),
            };
        }

        let body_text = String::from_utf8_lossy(&bytes).into_owned();
        return SafeFetchResult {
            ok: status.is_success(),
            status: status.as_u16(),
            final_url: current,
            body_text,
            error: None,
        };
    }

    SafeFetchResult {
        ok: false,
        status: 0,
        final_url: current,
        body_text: String::new(),
        error: Some("Too many redirects".into()),
    }
}

pub fn looks_like_closed_posting(html: &str, status: u16) -> bool {
    if status == 404 || status == 410 {
        return true;
    }
    let lower = html.to_lowercase();
    let closed_signals = [
        "no longer accepting applications",
        "job is closed",
        "this job has expired",
        "position has been filled",
        "this posting is no longer available",
        "sorry, this job is no longer available",
    ];
    closed_signals.iter().any(|s| lower.contains(s))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn closed_on_404() {
        assert!(looks_like_closed_posting("", 404));
        assert!(looks_like_closed_posting("", 410));
    }

    #[test]
    fn closed_copy() {
        assert!(looks_like_closed_posting(
            "<html>This job is closed. Thanks.</html>",
            200
        ));
    }

    #[tokio::test]
    async fn blocks_loopback() {
        let result = safe_fetch("http://127.0.0.1/", Some("GET"), None).await;
        assert!(!result.ok);
        assert!(result.error.as_deref().unwrap_or("").contains("Private"));
    }

    #[tokio::test]
    async fn blocks_local_hostname() {
        let result = safe_fetch("https://localhost/job", Some("GET"), None).await;
        assert!(!result.ok);
        assert!(result.error.as_deref().unwrap_or("").contains("local"));
    }

    #[tokio::test]
    async fn blocks_file_scheme() {
        let result = safe_fetch("file:///etc/passwd", Some("GET"), None).await;
        assert!(!result.ok);
        assert!(result.error.as_deref().unwrap_or("").contains("Only HTTP"));
    }

    #[tokio::test]
    async fn blocks_other_non_http_schemes() {
        let result = safe_fetch("ftp://example.com/job", Some("GET"), None).await;
        assert!(!result.ok);
        assert!(result.error.as_deref().unwrap_or("").contains("Only HTTP"));
    }
}
